snuba/query/expressions.py

```diff
@@ -0,0 +1,132 @@
+    function calls themselves in the AST), literals, etc.
+
+    All expressions can have an optional alias.
+    """
+
+    alias: Optional[str]
+
+    @abstractmethod
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        """
+        Transforms this expression through the function passed in input.
+        This works almost like a map function over sequences though, contrarily to
+        sequences, this acts on a subtree. The semantics of transform can be different
+        between intermediate nodes and leaves, so each node class can implement it
+        its own way.
+
+        All expressions are frozen dataclasses. This means they are immutable and
+        format will either return self or a new instance. It cannot transform the
+        expression in place.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def __iter__(self) -> Iterator[Expression]:
+        """
+        Used to iterate over this expression and its children. The exact
+        semantics depends on the structure of the expression.
+        See the implementations for more details.
+        """
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class Literal(Expression):
+    """
+    A literal in the SQL expression
+    """
+    value: Union[None, bool, str, float, int]
+
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        return func(self)
+
+    def __iter__(self) -> Iterator[Expression]:
+        yield self
+
+
+@dataclass(frozen=True)
+class Column(Expression):
+    """
+    Represent a column in the schema of the dataset.
+    """
+    column_name: str
+    table_name: Optional[str]
+
+    def transform(self, func: Callable[[Expression], Expression]) -> Expression:
+        return func(self)
+
+    def __iter__(self) -> Iterator[Expression]:
+        yield self
+
+
+@dataclass(frozen=True)
+class FunctionCall(Expression):
+    """
+    Represents an expression that resolves to a function call on Clickhouse.
+    This class also represent conditions. Since Clickhouse supports both the conventional
+    infix notation for condition and the functional one, we converge into one
+    representation only in the AST to make query processing easier.
+    A query processor would not have to care of processing both functional conditions
+    and infix conditions.
+
+    This function call abstraction can support both functions with one parameters group
+    like f(X,Y) and functions with two parameters groups f(X, Y)(Z).
+    Since we are not doing anything except transforming and iterating over these functions
+    there is no need to have a subclass for these two cases.
+    """
+    function_name: str
+    parameters_group1: Sequence[Expression]
+    # This is for expressions like f(x)(y).
+    # None means there is no second gorup of parameters
+    parameters_group2: Optional[Sequence[Expression]] = None
```

Comment:
Would it make sense to instead structure this as:

```
function: Union[str, FunctionCall]
parameters: Sequence[Expression]
```

… or is this too permissive, since this could lead to calls like `f(g(x)(y))(z)`? (Is that even a valid construct? I haven't tested it yet, so I honestly have no idea.)

Reply:
I think the issue with this is not the example you gave. That one is valid, but it was possible before as well by having one parameter of the first group of f, being a FunctionCall.
The issue is that I could express this:
```
f(x)(y)(z)(y)
```

_My functional programming enthusiast self speaks now:_
My interpretation of this syntax in Clickhouse is that that two sets of parameters are just useless syntactic sugar. But now that you mention it, this may actually a proper currying implementation, in which case it would make sense to consider the external call as an application of a function to the function returned by the application of the internal one.
z = f(x)(y) = f(x) returns g(k) -> z = g(y).
Which means f(x)(y)(z)(w) should be allowed. 
_My functional programming enthusiast self stops speaking now_

In practice my concern with expressing currying properly is that, while it is extremely powerful to evaluate expressions, it may make query processing harder. 
I think it is worth giving it a shot anyway, I may just split the FunctionCall implementation, instead of using the Union, this will make it clearer for query processors. 

```
# I need to think over naming better. But this is the idea.
class FunctionCall(ABC):

class BasicFunctionCall(FunctionCall):
    name: str
    params[...]

class CurriedFunctionCall(FunctionCall):
    internal: FunctionCall
     params[...]
```

What do you think?

Reply:
> I think the issue with this is not the example you gave. That one is valid, but it was possible before as well by having one parameter of the first group of f, being a FunctionCall.

Oops — yeah, that was a bad example, but based on the remainder of your comment I think you ended up on the same train of thought that I was.

> But now that you mention it, this may actually a proper currying implementation, in which case it would make sense to consider the external call as an application of a function to the function returned by the application of the internal one.

Yeah, this is how I've thought always of it — whether or not that's an accurate reflection of their intent, I'm not sure one way or another.

> In practice my concern with expressing currying properly is that, while it is extremely powerful to evaluate expressions, it may make query processing harder.

Do you have any examples why? Nothing sticks out to me immediately but I haven't put much thought into this.

> I think it is worth giving it a shot anyway, I may just split the FunctionCall implementation, instead of using the Union, this will make it clearer for query processors.
> 
> ```
> # I need to think over naming better. But this is the idea.
> class FunctionCall(ABC):
> 
> class BasicFunctionCall(FunctionCall):
>     name: str
>     params[...]
> 
> class CurriedFunctionCall(FunctionCall):
>     internal: FunctionCall
>      params[...]
> ```
> 
> What do you think?

I think this makes sense in general.

I don't think that we need a shared supertype for these — I'd consider just having `FunctionCall(Expression)` and `CurriedFunctionCall(Expression)` with the same attributes you described above. I think that not allowing infinitely recursive currying would be an advantage, since it's unclear whether or not ClickHouse allows that (and I don't think we have a use-case for it right now.)